File: app/internal/publisher/template/go/hardware.go
package wsd

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net"
	"os/exec"
	"runtime"
	"strings"
)

type HardwareDetector struct {
	fingerprint string
	identifiers map[string]string
}

func NewHardwareDetector() *HardwareDetector {
	return &HardwareDetector{}
}

func (h *HardwareDetector) GetFingerprint() string {
	if h.fingerprint != "" {
		return h.fingerprint
	}
	identifiers := h.collectIdentifiers()
	combined := h.buildCombinedString(identifiers)
	h.fingerprint = h.hashIdentifiers(combined)
	h.identifiers = identifiers
	return h.fingerprint
}

func (h *HardwareDetector) GetIdentifiers() map[string]string {
	if h.identifiers == nil {
		h.GetFingerprint()
	}
	result := make(map[string]string)
	for k, v := range h.identifiers {
		result[k] = v
	}
	return result
}

func (h *HardwareDetector) collectIdentifiers() map[string]string {
	identifiers := make(map[string]string)
	cpuID := h.getCPUID()
	if cpuID != "" {
		identifiers["cpu_id"] = cpuID
	}
	mbID := h.getMotherboardID()
	if mbID != "" {
		identifiers["motherboard_id"] = mbID
	}
	if mbID == "" {
		netID := h.getNetworkID()
		if netID != "" {
			identifiers["network_id"] = netID
		}
	}
	osInfo := h.getOSInfo()
	if osInfo != "" {
		identifiers["os_info"] = osInfo
	}
	return identifiers
}

func (h *HardwareDetector) getCPUID() string {
	switch runtime.GOOS {
	case "windows":
		out, err := exec.Command("wmic", "cpu", "get", "ProcessorId", "/value").Output()
		if err == nil {
			output := string(out)
			for _, line := range strings.Split(output, "\n") {
				line = strings.TrimSpace(line)
				if strings.HasPrefix(line, "ProcessorId=") {
					val := strings.TrimPrefix(line, "ProcessorId=")
					if val != "" {
						return val
					}
				}
			}
		}
	case "darwin":
		out, err := exec.Command("sysctl", "-n", "hw.model").Output()
		if err == nil {
			model := strings.TrimSpace(string(out))
			if model != "" {
				return "mac-" + model
			}
		}
		out, err = exec.Command("sysctl", "-n", "machdep.cpu.brand_string").Output()
		if err == nil {
			brand := strings.TrimSpace(string(out))
			if brand != "" {
				hash := sha256.Sum256([]byte(brand))
				return hex.EncodeToString(hash[:16])
			}
		}
	case "linux":
		out, err := exec.Command("sh", "-c", "cat /proc/cpuinfo").Output()
		if err == nil {
			content := string(out)
			for _, line := range strings.Split(content, "\n") {
				line = strings.TrimSpace(line)
				if strings.HasPrefix(line, "Serial") && strings.Contains(line, ":") {
					parts := strings.SplitN(line, ":", 2)
					val := strings.TrimSpace(parts[1])
					if val != "" {
						return "cpu-" + val
					}
				}
			}
			var vendor, family string
			for _, line := range strings.Split(content, "\n") {
				line = strings.TrimSpace(line)
				if strings.HasPrefix(line, "vendor_id") && strings.Contains(line, ":") {
					parts := strings.SplitN(line, ":", 2)
					vendor = strings.TrimSpace(parts[1])
				} else if strings.HasPrefix(line, "cpu family") && strings.Contains(line, ":") {
					parts := strings.SplitN(line, ":", 2)
					family = strings.TrimSpace(parts[1])
				}
			}
			if vendor != "" && family != "" {
				return vendor + "-" + family
			}
		}
	}
	return ""
}

func (h *HardwareDetector) getMotherboardID() string {
	switch runtime.GOOS {
	case "windows":
		out, err := exec.Command("wmic", "baseboard", "get", "SerialNumber", "/value").Output()
		if err == nil {
			output := string(out)
			for _, line := range strings.Split(output, "\n") {
				line = strings.TrimSpace(line)
				if strings.HasPrefix(line, "SerialNumber=") {
					val := strings.TrimPrefix(line, "SerialNumber=")
					val = strings.TrimSpace(val)
					if val != "" && val != "To be filled by O.E.M." && val != "Default string" {
						return "mb-" + val
					}
				}
			}
		}
	case "linux":
		out, err := exec.Command("dmidecode", "-s", "baseboard-serial-number").Output()
		if err == nil {
			serial := strings.TrimSpace(string(out))
			if serial != "" && serial != "To be filled by O.E.M." && serial != "Default string" {
				return "mb-" + serial
			}
		}
	}
	return ""
}

func (h *HardwareDetector) getNetworkID() string {
	interfaces, err := net.Interfaces()
	if err != nil {
		return ""
	}
	for _, iface := range interfaces {
		if iface.Flags&net.FlagLoopback == 0 && len(iface.HardwareAddr) > 0 {
			mac := iface.HardwareAddr.String()
			hash := sha256.Sum256([]byte("net-" + mac))
			return hex.EncodeToString(hash[:16])
		}
	}
	return ""
}

func (h *HardwareDetector) getOSInfo() string {
	return runtime.GOOS + "-" + runtime.GOARCH
}

func (h *HardwareDetector) buildCombinedString(identifiers map[string]string) string {
	parts := []string{}
	order := []string{"cpu_id", "motherboard_id", "network_id"}
	for _, key := range order {
		if val, ok := identifiers[key]; ok {
			parts = append(parts, val)
		}
	}
	return strings.Join(parts, "|")
}

func (h *HardwareDetector) hashIdentifiers(data string) string {
	hash := sha256.Sum256([]byte(data))
	return hex.EncodeToString(hash[:])
}

File: app/internal/publisher/template/rust/src/hardware.rs
use sha2::{Digest, Sha256};
use std::collections::HashMap;

pub struct HardwareDetector {
    fingerprint: Option<String>,
    identifiers: Option<HashMap<String, String>>,
}

impl HardwareDetector {
    pub fn new() -> Self {
        Self {
            fingerprint: None,
            identifiers: None,
        }
    }

    pub fn get_fingerprint(&mut self) -> String {
        if let Some(ref fp) = self.fingerprint {
            return fp.clone();
        }
        let identifiers = self.collect_identifiers();
        let combined = self.build_combined_string(&identifiers);
        let fp = self.hash_identifiers(&combined);
        self.fingerprint = Some(fp.clone());
        self.identifiers = Some(identifiers);
        fp
    }

    pub fn get_identifiers(&mut self) -> HashMap<String, String> {
        if self.identifiers.is_none() {
            self.get_fingerprint();
        }
        self.identifiers.clone().unwrap_or_default()
    }

    fn collect_identifiers(&self) -> HashMap<String, String> {
        let mut identifiers = HashMap::new();
        if let Some(cpu_id) = self.get_cpu_id() {
            identifiers.insert("cpu_id".to_string(), cpu_id);
        }
        if let Some(mb_id) = self.get_motherboard_id() {
            identifiers.insert("motherboard_id".to_string(), mb_id);
        }
        if !identifiers.contains_key("motherboard_id") {
            if let Some(net_id) = self.get_network_id() {
                identifiers.insert("network_id".to_string(), net_id);
            }
        }
        if let Some(os_info) = self.get_os_info() {
            identifiers.insert("os_info".to_string(), os_info);
        }
        identifiers
    }

    fn get_cpu_id(&self) -> Option<String> {
        #[cfg(target_os = "windows")]
        {
            if let Ok(out) = std::process::Command::new("wmic")
                .args(["cpu", "get", "ProcessorId", "/value"])
                .output()
            {
                let output = String::from_utf8_lossy(&out.stdout);
                for line in output.lines() {
                    let line = line.trim();
                    if let Some(val) = line.strip_prefix("ProcessorId=") {
                        let val = val.trim().to_string();
                        if !val.is_empty() {
                            return Some(val);
                        }
                    }
                }
            }
        }
        #[cfg(target_os = "macos")]
        {
            if let Ok(out) = std::process::Command::new("sysctl")
                .args(["-n", "hw.model"])
                .output()
            {
                let model = String::from_utf8_lossy(&out.stdout).trim().to_string();
                if !model.is_empty() {
                    return Some(format!("mac-{}", model));
                }
            }
            if let Ok(out) = std::process::Command::new("sysctl")
                .args(["-n", "machdep.cpu.brand_string"])
                .output()
            {
                let brand = String::from_utf8_lossy(&out.stdout).trim().to_string();
                if !brand.is_empty() {
                    let mut hasher = Sha256::new();
                    hasher.update(brand.as_bytes());
                    let hash = hex::encode(hasher.finalize());
                    return Some(hash[..16].to_string());
                }
            }
        }
        #[cfg(target_os = "linux")]
        {
            if let Ok(content) = std::fs::read_to_string("/proc/cpuinfo") {
                for line in content.lines() {
                    let line = line.trim();
                    if let Some(val) = line.strip_prefix("Serial").and_then(|s| s.split(':').nth(1)) {
                        let val = val.trim();
                        if !val.is_empty() {
                            return Some(format!("cpu-{}", val));
                        }
                    }
                }
                let mut vendor = String::new();
                let mut family = String::new();
                for line in content.lines() {
                    let line = line.trim();
                    if line.starts_with("vendor_id") {
                        if let Some(val) = line.split(':').nth(1) {
                            vendor = val.trim().to_string();
                        }
                    } else if line.starts_with("cpu family") {
                        if let Some(val) = line.split(':').nth(1) {
                            family = val.trim().to_string();
                        }
                    }
                }
                if !vendor.is_empty() && !family.is_empty() {
                    return Some(format!("{}-{}", vendor, family));
                }
            }
        }
        None
    }

    fn get_motherboard_id(&self) -> Option<String> {
        #[cfg(target_os = "windows")]
        {
            if let Ok(out) = std::process::Command::new("wmic")
                .args(["baseboard", "get", "SerialNumber", "/value"])
                .output()
            {
                let output = String::from_utf8_lossy(&out.stdout);
                for line in output.lines() {
                    let line = line.trim();
                    if let Some(val) = line.strip_prefix("SerialNumber=") {
                        let val = val.trim().to_string();
                        if !val.is_empty()
                            && val != "To be filled by O.E.M."
                            && val != "Default string"
                        {
                            return Some(format!("mb-{}", val));
                        }
                    }
                }
            }
        }
        #[cfg(target_os = "linux")]
        {
            if let Ok(out) = std::process::Command::new("dmidecode")
                .args(["-s", "baseboard-serial-number"])
                .output()
            {
                let serial = String::from_utf8_lossy(&out.stdout).trim().to_string();
                if !serial.is_empty()
                    && serial != "To be filled by O.E.M."
                    && serial != "Default string"
                {
                    return Some(format!("mb-{}", serial));
                }
            }
        }
        None
    }

    fn get_network_id(&self) -> Option<String> {
        if let Ok(Some(mac)) = mac_address::get_mac_address() {
            let mac_str = mac.to_string();
            let mut hasher = Sha256::new();
            hasher.update(format!("net-{}", mac_str).as_bytes());
            let hash = hex::encode(hasher.finalize());
            return Some(hash[..16].to_string());
        }
        None
    }

    fn get_os_info(&self) -> Option<String> {
        Some(format!(
            "{}-{}",
            std::env::consts::OS,
            std::env::consts::ARCH
        ))
    }

    fn build_combined_string(&self, identifiers: &HashMap<String, String>) -> String {
        let order = ["cpu_id", "motherboard_id", "network_id"];
        let parts: Vec<&str> = order
            .iter()
            .filter_map(|k| identifiers.get(*k).map(|s| s.as_str()))
            .collect();
        parts.join("|")
    }

    fn hash_identifiers(&self, data: &str) -> String {
        let mut hasher = Sha256::new();
        hasher.update(data.as_bytes());
        hex::encode(hasher.finalize())
    }
}

impl Default for HardwareDetector {
    fn default() -> Self {
        Self::new()
    }
}
